repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java

```diff
@@ -454,6 +453,15 @@ private boolean incrementAndCheckIfRelationsLimitReached(AtlasEdge atlasEdge, bo
         String                     outGuid                   = AtlasGraphUtilsV2.getIdFromVertex(outVertex);
         LineageOnDemandConstraints outGuidLineageConstraints = getAndValidateLineageConstraintsByGuid(outGuid, atlasLineageOnDemandContext);
 
+        // Special checks to handle self cyclic lineage on base entity
+        boolean areBothDirectionsRequested = baseEntityHasSelfCyclicLineage && AtlasLineageOnDemandInfo.LineageDirection.BOTH.equals(outGuidLineageConstraints.getDirection());
+        boolean skipOutputIncrement = baseEntityHasSelfCyclicLineage && isInput && areBothDirectionsRequested;
+        boolean skipVisitedEdgeCheck = baseEntityHasSelfCyclicLineage && !isInput && areBothDirectionsRequested;
+
+        if (!skipVisitedEdgeCheck && lineageContainsVisitedEdgeV2(ret, atlasEdge)) {
```

Comment:
does not cover cyclic cases but it was not the target I assume since we will be looking at it later